services/slam/fake/slam.go

```diff
@@ -0,0 +1,160 @@
+// Package fake implements a fake base.
+package fake
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"image"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"go.viam.com/utils"
+	"go.viam.com/utils/artifact"
+
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage"
+	"go.viam.com/rdk/services/slam"
+	"go.viam.com/rdk/spatialmath"
+	rdkutils "go.viam.com/rdk/utils"
+	"go.viam.com/rdk/vision"
+)
+
+var model = resource.NewDefaultModel("fake")
+
+const (
+	internalStateTemplate = "slam/example_cartographer_outputs/internal_state/internal_state_%d.pbstream"
+	maxDataCount          = 16
+	pcdTemplate           = "slam/example_cartographer_outputs/pointcloud/pointcloud_%d.pcd"
+	pngTemplate           = "slam/example_cartographer_outputs/image_map/image_map_%d.png"
+	positionTemplate      = "slam/example_cartographer_outputs/position/position_%d.txt"
+)
+
+func init() {
+	registry.RegisterService(
+		slam.Subtype,
+		model,
+		registry.Service{
+			Constructor: func(
+				ctx context.Context,
+				_ registry.Dependencies,
+				config config.Service,
+				logger golog.Logger,
+			) (interface{}, error) {
+				return &SLAM{Name: config.Name, dataCount: 1}, nil
+			},
+		},
+	)
+}
+
+var _ = slam.Service(&SLAM{})
+
+// SLAM is a fake slam that returns generic data.
+type SLAM struct {
+	generic.Echo
+	Name      string
+	dataCount int
+}
+
+func (slamSvc *SLAM) GetMap(ctx context.Context, name, mimeType string, cp *referenceframe.PoseInFrame,
+	include bool, extra map[string]interface{},
+) (string, image.Image, *vision.Object, error) {
+	var err error
+	var img image.Image
+	var vObj *vision.Object
```

Comment:
Will `img` and `vObj` both be nil here initially?

Reply:
Short answer is yes.

More specifically:
`img` is `nil` here
`vObj` is `(*vision.Object)(nil)`, which b/c it is not an interface value, `pcData == nil` returns `true`, which is good.

B/c one of the return values of `GetMap` is `*vision.Object`, even in the early returns where it is explicitly set to `nil` in server.go, the vision object pointer will still be `(*vision.Object)(nil)`
 
This is following the same pattern in builtin.go https://github.com/viamrobotics/rdk/blob/main/services/slam/builtin/builtin.go#L499

Reply:
thank you!